File: desktop/test/merge-rules.test.ts
// Részleges szabályok a szinkronban.
//
// Két kimenetel van, ami rosszabb, mint ha a szabályok egyáltalán nem
// szinkronizálódnának:
//
//   1. egy szabály CSENDBEN eltűnik (a felhasználó azt hiszi, tilt, és nem);
//   2. egy kifizetett eltávolítás visszajön (a próbatétel értéktelen lesz).
//
// A legalattomosabb az első egy változata: egy RÉGI app-verzió nem ismeri a
// mezőt, tehát ami átmegy rajta, abból eltűnik. Ha ezt „mindent töröltek”-ként
// értenénk, elég lenne egy frissítetlen telefon a fiókban.

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import { mergeSite, mergeSiteLists, type SyncSite } from '../src/shared/sync/merge';
import { normalizeRule } from '../src/shared/urlrules';

const R = (s: string) => normalizeRule(s)!;

function site(extra: Partial<SyncSite> = {}): SyncSite {
  return {
    id: 'site_1', domain: 'youtube.com', hostnames: ['youtube.com'], addedAt: 1000,
    pauseUntil: null, pendingDeleteAt: null,
    rev: 1, updatedAt: 1000, updatedBy: 'gep-a',
    ...extra,
  };
}

const labels = (s: SyncSite) => (s.rules ?? []).map((r) => `${r.host}${r.path}`).sort();

test('rules added on two devices at once are both kept', () => {
  // Egyenlő rev: senki nem „újabb”. Ha ilyenkor egy egész listát választanánk,
  // az egyik eszközön felvett szabály némán elveszne — a felhasználó pedig azt
  // hinné, hogy felvette.
  const a = site({ rev: 5, rules: [R('youtube.com/@egy')] });
  const b = site({ rev: 5, rules: [R('youtube.com/@ketto')], updatedBy: 'gep-b' });
  assert.deepEqual(labels(mergeSite(a, b)), ['youtube.com/@egy', 'youtube.com/@ketto']);
  // Szimmetrikus: minden eszköz ugyanarra jut, különben örökké írnák egymást.
  assert.deepEqual(labels(mergeSite(b, a)), labels(mergeSite(a, b)));
});

test('a removal that was paid for is not resurrected', () => {
  // Az eltávolítás próbatételbe kerül, és a `rev`-et lépteti. A nagyobb rev
  // mögött ott a munka: az ő listája érvényes, nem az egyesítés.
  const before = site({ rev: 5, rules: [R('youtube.com/@egy'), R('youtube.com/@ketto')] });
  const after = site({ rev: 6, rules: [R('youtube.com/@ketto')], updatedAt: 2000 });
  assert.deepEqual(labels(mergeSite(before, after)), ['youtube.com/@ketto']);
  assert.deepEqual(labels(mergeSite(after, before)), ['youtube.com/@ketto']);

  // És az utolsó szabály levétele sem jön vissza.
  const empty = site({ rev: 7, rules: [], updatedAt: 3000 });
  assert.deepEqual(labels(mergeSite(after, empty)), []);
});

test('an app version that does not know the field cannot delete the rules', () => {
  // EZ A LEGVESZÉLYESEBB ESET. Egy régi kliens a mezőt nem érti, tehát ami
  // átmegy rajta, abból hiányzik. Ha a hiányt „mindent töröltek”-nek vennénk,
  // elég lenne egy frissítetlen telefon a fiókban, és a gépen felvett összes
  // szabály csendben eltűnne — próbatétel nélkül, jelzés nélkül.
  const mine = site({ rev: 5, rules: [R('youtube.com/@egy')] });
  const old = site({ rev: 9, updatedAt: 9000, updatedBy: 'regi-telefon' }); // nincs `rules` kulcs
  assert.deepEqual(labels(mergeSite(mine, old)), ['youtube.com/@egy'],
    'a nagyobb rev sem törölhet olyan mezőt, amiről nem tud');
  assert.deepEqual(labels(mergeSite(old, mine)), ['youtube.com/@egy']);

  // Az ÜRES LISTA viszont valódi állítás: „volt, és levettem”.
  const emptied = site({ rev: 9, rules: [], updatedAt: 9000 });
  assert.deepEqual(labels(mergeSite(mine, emptied)), []);
});

test('a site that never had rules stays without the field', () => {
  // Ha minden rekordba beletennénk egy üres tömböt, két szerkezetileg azonos
  // lista különbözőnek látszana, és a szinkron minden körben feltöltene.
  const a = site({ rev: 2 });
  const b = site({ rev: 3, updatedAt: 2000 });
  assert.equal(Object.prototype.hasOwnProperty.call(mergeSite(a, b), 'rules'), false);
});

test('junk from the other device does not become a rule', () => {
  // A szinkronon át érkező adat ugyanolyan megbízhatatlan, mint bármi más. Egy
  // út nélküli „szabály” az EGÉSZ oldalt jelentené a bővítményben — vagyis a
  // gyengébb réteg többet tiltana, mint amit bárki beállított.
  const a = site({
    rev: 5,
    rules: [
      { host: 'youtube.com', path: '' },                    // nincs út
      { host: '', path: '/@valaki' },                       // nincs hoszt
      { host: 'youtube.com', path: '/@ok' },
      { host: 'youtube.com', path: '/@ok' },                // duplikátum
      { host: 'M.YouTube.com', path: '/@Masik' },           // mobil hoszt, nagybetű
      null as unknown as { host: string; path: string },
    ],
  });
  const b = site({ rev: 5, updatedBy: 'gep-b' });
  assert.deepEqual(labels(mergeSite(a, b)), ['youtube.com/@masik', 'youtube.com/@ok']);
});

test('the rule list cannot grow without bound through sync', () => {
  // Két eszköz ötven-ötven szabállyal: az egyesítés száz lenne. A korlát nem
  // szépészet — a lista minden körben felmegy és lejön, titkosítva.
  const many = (prefix: string) => Array.from({ length: 50 }, (_, i) => R(`youtube.com/@${prefix}${i}`));
  const a = site({ rev: 5, rules: many('a') });
  const b = site({ rev: 5, rules: many('b'), updatedBy: 'gep-b' });
  assert.equal(mergeSite(a, b).rules!.length, 50);
});

test('rules survive a whole-list merge, and the union is stable', () => {
  const a: SyncSite[] = [site({ rev: 4, rules: [R('youtube.com/@egy')] })];
  const b: SyncSite[] = [site({ rev: 4, rules: [R('youtube.com/@ketto')], updatedBy: 'gep-b' })];
  const once = mergeSiteLists(a, b);
  assert.deepEqual(labels(once[0]), ['youtube.com/@egy', 'youtube.com/@ketto']);
  // Kétszer lefuttatva ugyanaz: enélkül a két eszköz felváltva írná felül
  // egymást, és a szinkron sosem érne véget.
  assert.deepEqual(mergeSiteLists(once, b), once);
  assert.deepEqual(mergeSiteLists(b, once), once);
});

File: android/jvm-tests/src/test/kotlin/MergeRulesTest.kt
import hu.breaker.app.core.SyncMerge
import hu.breaker.app.core.SyncMerge.SyncSite
import hu.breaker.app.core.UrlRules
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Részleges szabályok a szinkronban — a `desktop/test/merge-rules.test.ts` párja.
 *
 * Két kimenetel van, ami rosszabb, mint ha a szabályok egyáltalán nem
 * szinkronizálódnának:
 *
 *   1. egy szabály CSENDBEN eltűnik (a felhasználó azt hiszi, tilt, és nem);
 *   2. egy kifizetett eltávolítás visszajön (a próbatétel értéktelen lesz).
 *
 * A legalattomosabb az első egy változata: EZ AZ APP maga a „régi kliens”, ha
 * nem tud a mezőről. Androidon a szabályokat semmi nem érvényesíti, tárolni és
 * továbbadni viszont KELL őket — enélkül elég egy telefon a fiókban, és a gépen
 * felvett szabályok minden körben eltűnnének.
 */
class MergeRulesTest {

    private fun r(s: String) = UrlRules.normalizeRule(s)!!

    private fun site(
        rev: Int = 1,
        rules: List<UrlRules.UrlRule>? = null,
        updatedAt: Long = 1_000,
        updatedBy: String = "gep-a",
    ) = SyncSite(
        id = "site_1", domain = "youtube.com", hostnames = listOf("youtube.com"),
        addedAt = 1_000, pendingDeleteAt = null, schedule = null, dailyLimitSeconds = null,
        alias = null, rules = rules, rev = rev, updatedAt = updatedAt, updatedBy = updatedBy,
    )

    private fun labels(s: SyncSite) = (s.rules ?: emptyList()).map { it.host + it.path }.sorted()

    @Test fun `rules added on two devices at once are both kept`() {
        // Egyenlő rev: senki nem „újabb”. Ha ilyenkor egy egész listát
        // választanánk, az egyik eszközön felvett szabály némán elveszne.
        val a = site(rev = 5, rules = listOf(r("youtube.com/@egy")))
        val b = site(rev = 5, rules = listOf(r("youtube.com/@ketto")), updatedBy = "telefon")
        assertEquals(listOf("youtube.com/@egy", "youtube.com/@ketto"), labels(SyncMerge.mergeSite(a, b)))
        // Szimmetrikus: minden eszköz ugyanarra jut, különben örökké írnák egymást.
        assertEquals(labels(SyncMerge.mergeSite(a, b)), labels(SyncMerge.mergeSite(b, a)))
    }

    @Test fun `a removal that was paid for is not resurrected`() {
        val before = site(rev = 5, rules = listOf(r("youtube.com/@egy"), r("youtube.com/@ketto")))
        val after = site(rev = 6, rules = listOf(r("youtube.com/@ketto")), updatedAt = 2_000)
        assertEquals(listOf("youtube.com/@ketto"), labels(SyncMerge.mergeSite(before, after)))
        assertEquals(listOf("youtube.com/@ketto"), labels(SyncMerge.mergeSite(after, before)))

        val empty = site(rev = 7, rules = emptyList(), updatedAt = 3_000)
        assertEquals(emptyList(), labels(SyncMerge.mergeSite(after, empty)))
    }

    @Test fun `an app version that does not know the field cannot delete the rules`() {
        // EZ A LEGVESZÉLYESEBB ESET, és Androidon a legvalószínűbb: a szabályokat
        // itt semmi nem érvényesíti, tehát könnyű lenne „nem foglalkozni velük”.
        val mine = site(rev = 5, rules = listOf(r("youtube.com/@egy")))
        val old = site(rev = 9, rules = null, updatedAt = 9_000, updatedBy = "regi")
        assertEquals(listOf("youtube.com/@egy"), labels(SyncMerge.mergeSite(mine, old)),
            "a nagyobb rev sem törölhet olyan mezőt, amiről nem tud")
        assertEquals(listOf("youtube.com/@egy"), labels(SyncMerge.mergeSite(old, mine)))

        // Az ÜRES LISTA viszont valódi állítás: „volt, és levettem”.
        val emptied = site(rev = 9, rules = emptyList(), updatedAt = 9_000)
        assertEquals(emptyList(), labels(SyncMerge.mergeSite(mine, emptied)))
    }

    @Test fun `a site that never had rules stays without the field`() {
        val a = site(rev = 2)
        val b = site(rev = 3, updatedAt = 2_000)
        assertNull(SyncMerge.mergeSite(a, b).rules)
    }

    @Test fun `junk from the other device does not become a rule`() {
        // Egy út nélküli „szabály” az EGÉSZ oldalt jelentené a bővítményben —
        // vagyis a gyengébb réteg többet tiltana, mint amit bárki beállított.
        val a = site(rev = 5, rules = listOf(
            UrlRules.UrlRule("youtube.com", ""),
            UrlRules.UrlRule("", "/@valaki"),
            UrlRules.UrlRule("youtube.com", "/@ok"),
            UrlRules.UrlRule("youtube.com", "/@ok"),
            UrlRules.UrlRule("M.YouTube.com", "/@Masik"),
        ))
        val b = site(rev = 5, updatedBy = "telefon")
        assertEquals(listOf("youtube.com/@masik", "youtube.com/@ok"), labels(SyncMerge.mergeSite(a, b)))
    }

    @Test fun `the rule list cannot grow without bound through sync`() {
        fun many(prefix: String) = (0 until 50).map { r("youtube.com/@$prefix$it") }
        val a = site(rev = 5, rules = many("a"))
        val b = site(rev = 5, rules = many("b"), updatedBy = "telefon")
        assertEquals(UrlRules.MAX_RULES_PER_SITE, SyncMerge.mergeSite(a, b).rules!!.size)
    }

    @Test fun `rules survive a whole-list merge, and the union is stable`() {
        val a = listOf(site(rev = 4, rules = listOf(r("youtube.com/@egy"))))
        val b = listOf(site(rev = 4, rules = listOf(r("youtube.com/@ketto")), updatedBy = "telefon"))
        val once = SyncMerge.mergeLists(a, b)
        assertEquals(listOf("youtube.com/@egy", "youtube.com/@ketto"), labels(once[0]))
        // Kétszer lefuttatva ugyanaz: enélkül a két eszköz felváltva írná felül
        // egymást, és a szinkron sosem érne véget.
        assertEquals(once, SyncMerge.mergeLists(once, b))
        assertEquals(once, SyncMerge.mergeLists(b, once))
    }

    @Test fun `the uploaded JSON keeps the difference between unknown and emptied`() {
        // Ez a különbség a szinkron DRÓTFORMÁJÁN is meg kell maradjon: ha az
        // üres lista és a hiányzó kulcs ugyanúgy nézne ki, a fenti védelem a
        // hálózaton veszne el.
        val unknown = site(rev = 1, rules = null)
        val emptied = site(rev = 1, rules = emptyList())
        val withOne = site(rev = 1, rules = listOf(r("youtube.com/@egy")))

        assertTrue(!SyncClientJson.encode(unknown).contains("\"rules\""), "nincs kulcs, ha nem tudunk róla")
        assertTrue(SyncClientJson.encode(emptied).contains("\"rules\":[]"), "üres lista viszont kimegy")

        assertNull(SyncClientJson.decode(SyncClientJson.encode(unknown)).rules)
        assertEquals(emptyList(), SyncClientJson.decode(SyncClientJson.encode(emptied)).rules)
        assertEquals(listOf("youtube.com/@egy"), labels(SyncClientJson.decode(SyncClientJson.encode(withOne))))
    }
}

/** A `sitesToJson`/`sitesFromJson` egyetlen rekordra, hogy a teszt olvasható maradjon. */
private object SyncClientJson {
    fun encode(s: SyncSite): String =
        hu.breaker.app.core.SyncClient.sitesToJson(listOf(s)).replace(" ", "")

    fun decode(text: String): SyncSite = hu.breaker.app.core.SyncClient.sitesFromJson(text).first()
}
